doc/connectivity/networking/api/tls_credentials_shell.rst

```diff
@@ -0,0 +1,243 @@
+   "``<TYPE>``", "Optional. If provided, only list credentials with this credential type. Pass ``any`` or omit to allow any credential type. Otherwise, see :ref:`tls_credentials_shell_cred_types` for valid values."
+
+
+Output
+------
+
+The command outputs all matching credentials in the following (CSV-compliant) format:
+
+.. code-block:: shell
+
+   <SECTAG>,<TYPE>,<DIGEST>,<STATUS>
+
+Where:
+
+.. csv-table::
+   :header: "Symbol", "Value"
+   :widths: 15 85
+
+   "``<SECTAG>``", "The sectag of the listed credential. A non-negative integer."
+   "``<TYPE>``", "Credential type short-code (see :ref:`tls_credentials_shell_cred_types` for details) of the listed credential."
+   "``<DIGEST>``", "A string digest representing the credential contents. The exact nature of this digest may vary depending on credentials storage backend, but currently for all backends this is a base64 encoded SHA256 hash of the raw credential contents (so different storage formats for essentially identical credentials will have different digests)."
+   "``<STATUS>``", "Status code indicating success or failure with generating a digest of the listed credential. 0 if successful, negative error code specific to the storage backend otherwise. Lines for which status is not zero will be printed with error formatting."
+
+After the list is printed, a final summary of the found credentials will be printed in the form:
+
+.. code-block:: shell
+
+   <N> credentials found.
+
+Where `<N>` is the number of credentials found, and is zero if none are found.
+
+.. _tls_credentials_shell_cred_types:
+
+Credential Types
+****************
+
+The following keywords (case-insensitive) may be used to specify a credential type:
+
+.. csv-table::
+   :header: "Keyword(s)", "Meaning"
+   :widths: 15 85
+
+   "``CA_CERT``, ``CA``", "A trusted CA certificate."
+   "``SERVER_CERT``, ``SELF_CERT``, ``SELF``, ``SERV``", "Self or server certificate."
+   "``PRIVATE_KEY``, ``PK``", "A private key."
+   "``PRE_SHARED_KEY``, ``PSK``", "A pre-shared key."
+   "``PRE_SHARED_KEY_ID``, ``PSK_ID``", "ID for pre-shared key."
+
+.. _tls_credentials_shell_data_formats:
+
+Storage/Retrieval Formats
+*************************
+
+The following keywords (case-insensitive) specify how credential data is stored and retrieved.
+
+
+.. csv-table::
+   :header: "Keyword", "Meaning", "Behavior during storage (``cred add``)", "Behavior during retrieval (``cred get``)"
+   :widths: 3, 15, 41, 41
+
+   "``BIN``", "Credential is treated as non-terminated base64.", "Provided data will be decoded from base64 into raw binary before storage. No terminator will be appended.", "Stored data will be encoded into base64 before being printed."
+   "``BINT``", "Credential is treated as NULL-terminated base64.", "Provided data will be decoded from base64 into raw binary and a NULL terminator will be appended before storage.", "NULL terminator will be truncated from stored data before said data is encoded into base64 and then printed."
```

Comment:
Do you actually know of any real world situation that this is necessary?

It would make more sense to have a `B64` option that writes and reads the credential through the shell in base64 and also stores in base64.  A null terminator might be stored in this case if the underlying storage system does not have an associated length field.

Reply:
Can you clarify? When you say "reads the credential through shell in base64 and also stores in base64", this sounds to me like the text buffer written by the user would just be passed directly to the tls_credentials library, without a NULL terminator -- but that's exactly what the `strt` option does

If, instead, you mean the text buffer would be parsed from base64 into the actual binary that base64 represents, and then that **_binary_** would be passed directly to the tls_credentials library, then that is precisely what the `bin` mode does.

That's why I avoided `b64`, it makes it sound like we're storing raw base64. `bin` points to the fact that the data being entered gets converted to raw binary, whereas `str` interprets the provided data as a verbatim string.

As for whether there is a real world situation that `bint` is necessary, `bin` is the only mode that is truly necessary, the rest are for convenience. See discussion here:

https://github.com/zephyrproject-rtos/zephyr/pull/64343#discussion_r1372178912

Reply:
Oh and to further clarify, the storage backend always has an associated length field, but mbedTLS requires certain cert formats to include a NULL terminator. In fact it uses the length field to check whether the last character is NULL as part of its process for determining what format was used